crypto/key.go

```diff
@@ -0,0 +1,205 @@
+	Crypto CipherJSON
+}
+
+func (k *Key) Address() []byte {
+	pubBytes := FromECDSAPub(&k.PrivateKey.PublicKey)
+	return Sha3(pubBytes)[12:]
+}
+
+func (k *Key) MarshalJSON() (j []byte, err error) {
+	stringStruct := PlainKeyJSON{
+		k.Id.String(),
+		hex.EncodeToString(k.Flags[:]),
+		hex.EncodeToString(FromECDSA(k.PrivateKey)),
+	}
+	j, err = json.Marshal(stringStruct)
+	return j, err
+}
+
+func (k *Key) UnmarshalJSON(j []byte) (err error) {
+	keyJSON := new(PlainKeyJSON)
+	err = json.Unmarshal(j, &keyJSON)
+	if err != nil {
+		return err
+	}
+
+	u := new(uuid.UUID)
+	*u = uuid.Parse(keyJSON.Id)
+	if *u == nil {
+		err = errors.New("UUID parsing failed")
+		return err
+	}
+	k.Id = u
+
+	flagsBytes, err := hex.DecodeString(keyJSON.Flags)
+	if err != nil {
+		return err
+	}
+
+	PrivateKeyBytes, err := hex.DecodeString(keyJSON.PrivateKey)
+	if err != nil {
+		return err
+	}
+
+	copy(k.Flags[:], flagsBytes[0:4])
+	k.PrivateKey = ToECDSA(PrivateKeyBytes)
+
+	return err
+}
+
+func NewKey() *Key {
+	randBytes := GetEntropyCSPRNG(32)
+	reader := bytes.NewReader(randBytes)
+	_, x, y, err := elliptic.GenerateKey(S256(), reader)
+	if err != nil {
+		panic("key generation: elliptic.GenerateKey failed: " + err.Error())
+	}
+	privateKeyMarshalled := elliptic.Marshal(S256(), x, y)
+	privateKeyECDSA := ToECDSA(privateKeyMarshalled)
+
+	key := new(Key)
+	id := uuid.NewRandom()
+	key.Id = &id
+	// flags := new([4]byte)
+	// key.Flags = flags
+	key.PrivateKey = privateKeyECDSA
+	return key
+}
+
+// plain crypto/rand. this is /dev/urandom on Unix-like systems.
+func GetEntropyCSPRNG(n int) []byte {
+	mainBuff := make([]byte, n)
+	_, err := io.ReadFull(crand.Reader, mainBuff)
+	if err != nil {
+		panic("key generation: reading from crypto/rand failed: " + err.Error())
+	}
+	return mainBuff
+}
+
+// TODO: verify. Do not use until properly discussed.
+// we start with crypt/rand, then mix in additional sources of entropy.
+// These sources are from three types: OS, go runtime and ethereum client state.
+func GetEntropyTinFoilHat() []byte {
+	startTime := time.Now().UnixNano()
+	// for each source, we XOR in it's SHA3 hash.
+	mainBuff := GetEntropyCSPRNG(32)
+	// 1. OS entropy sources
+	startTimeBytes := make([]byte, 32)
+	binary.PutVarint(startTimeBytes, startTime)
+	startTimeHash := Sha3(startTimeBytes)
+	mix32Byte(mainBuff, startTimeHash)
+
+	pid := os.Getpid()
+	pidBytes := make([]byte, 32)
+	binary.PutUvarint(pidBytes, uint64(pid))
+	pidHash := Sha3(pidBytes)
+	mix32Byte(mainBuff, pidHash)
+
+	osEnv := os.Environ()
+	osEnvBytes := []byte(strings.Join(osEnv, ""))
+	osEnvHash := Sha3(osEnvBytes)
+	mix32Byte(mainBuff, osEnvHash)
+
+	// not all OS have hostname in env variables
+	osHostName, err := os.Hostname()
+	if err != nil {
+		osHostNameBytes := []byte(osHostName)
+		osHostNameHash := Sha3(osHostNameBytes)
+		mix32Byte(mainBuff, osHostNameHash)
+	}
+
+	// 2. go runtime entropy sources
+	memStats := new(runtime.MemStats)
+	runtime.ReadMemStats(memStats)
```

Comment:
Maybe you can use [`ReadGCStats`](http://golang.org/pkg/runtime/debug/#ReadGCStats) from `runtime/debug`.